### PR TITLE
feature/menu

### LINKED ISSUE #20: Add a demo for the Menu Component

### PR DESCRIPTION
**Pull request by DaniloDamico**
Adds #20 from scratch.

This commit implements the Menu Component.
More specifically, it contains the demo of an overflow menu, a dropdown menu and an exposed dropdown menu

### CODE REVIEW

File: app/src/main/res/drawable/ic_info_black_24dp.xml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+<vector xmlns:android="http://schemas.android.com/apk/res/android"
```
Comment (buracchi) [inline app/src/main/res/drawable/ic_info_black_24dp.xml]: Simple vector images can change color at runtime, remove the color from their name.

File: app/src/main/res/layout/activity_menu.xml [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+
+    <include
+        android:id="@+id/app_bar_container"
+        layout="@layout/app_bar_demo" />
+
+    <androidx.core.widget.NestedScrollView
+        android:id="@+id/menu_scroll_view"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:layout_marginStart="@dimen/margin_medium"
```
Comment (buracchi) [inline app/src/main/res/layout/activity_menu.xml]: Don't add extra margins

File: app/src/main/res/layout/activity_menu.xml [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        android:layout_marginEnd="@dimen/margin_medium"
+        app:layout_behavior="com.google.android.material.appbar.AppBarLayout$ScrollingViewBehavior">
+
+        <LinearLayout
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:layout_marginTop="@dimen/margin_medium"
+            android:orientation="vertical">
```
Comment (buracchi) [inline app/src/main/res/layout/activity_menu.xml:28]: Don't delete the existing margins

File: app/src/main/res/layout/activity_menu.xml [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+            <TextView
+                android:id="@+id/text_view_span_title"
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:text="Title"
+                android:layout_margin="@dimen/margin_medium"
```
Comment (buracchi) [inline app/src/main/res/layout/activity_menu.xml]: Remove the margins from the span text view

File: app/src/main/res/layout/activity_menu.xml [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content">
+
+                <TextView
+                    android:id="@+id/text_view_span_dropdown"
+                    android:layout_width="match_parent"
+                    android:layout_height="wrap_content"
+                    android:layout_margin="@dimen/margin_medium"
+                    android:text="Dropdown"
+                    app:layout_constraintStart_toStartOf="parent"
```
Comment (buracchi) [inline app/src/main/res/layout/activity_menu.xml]: The toStartOf costraint is useless here

File: app/src/main/res/layout/activity_menu.xml [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+                    android:id="@+id/btn_dropdown_menu"
+                    android:layout_width="wrap_content"
+                    android:layout_height="wrap_content"
+                    android:hint="Dropdown Menu"
+                    android:onClick="showPopupMenu"
+                    app:layout_constraintBottom_toBottomOf="parent"
+                    app:layout_constraintEnd_toEndOf="parent"
+                    app:layout_constraintStart_toStartOf="parent"
+                    app:layout_constraintTop_toBottomOf="@+id/text_view_span_dropdown"
+                    tools:ignore="HardcodedText" />
```
Comment (buracchi) [inline app/src/main/res/layout/activity_menu.xml:84]: Do you seriously think hard encoding text in a button is not a silly idea?

File: app/src/main/res/layout/activity_menu.xml [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+                    app:layout_constraintStart_toStartOf="parent"
+                    app:layout_constraintTop_toBottomOf="@+id/text_view_span_exposed_dropdown"
+                    tools:ignore="HardcodedText">
+
+                    <AutoCompleteTextView
+                        android:id="@+id/filled_exposed_dropdown"
+                        android:layout_width="match_parent"
+                        android:layout_height="wrap_content"
+
+                        app:layout_constraintVertical_bias="0.421"
```
Comment (buracchi) [inline app/src/main/res/layout/activity_menu.xml]: I'm quite sure you don't want to do this.

File: app/src/main/res/layout/activity_menu.xml [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+                <com.google.android.material.textfield.TextInputLayout
+                    style="@style/Widget.MaterialComponents.TextInputLayout.FilledBox.ExposedDropdownMenu"
+                    android:layout_width="0dp"
+                    android:layout_height="wrap_content"
+                    android:hint="Exposed dropdown"
+                    android:layout_marginStart="@dimen/margin_medium"
+                    android:layout_marginEnd="@dimen/margin_medium"
+                    app:layout_constraintBottom_toBottomOf="parent"
+                    app:layout_constraintEnd_toEndOf="parent"
+                    app:layout_constraintStart_toStartOf="parent"
```
Comment (buracchi) [inline app/src/main/res/layout/activity_menu.xml:112]: Add a large marginBottom

File: app/src/main/res/layout/activity_menu.xml [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+                    android:id="@+id/text_view_span_exposed_dropdown"
+                    android:layout_width="match_parent"
+                    android:layout_height="wrap_content"
+                    android:layout_margin="@dimen/margin_medium"
+                    android:text="Exposed Dropdown"
+                    app:layout_constraintStart_toStartOf="parent"
+                    app:layout_constraintTop_toTopOf="parent"
+                    tools:ignore="HardcodedText" />
+
+                <com.google.android.material.textfield.TextInputLayout
```
Comment (buracchi) [inline app/src/main/res/layout/activity_menu.xml:102]: Hints are covered by the keyboard, fix it.

File: app/src/main/res/layout/menu_exposed_menu_item.xml [hunk 10]
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="utf-8"?>
+<TextView xmlns:android="http://schemas.android.com/apk/res/android"
+    android:layout_width="match_parent"
+    android:layout_height="wrap_content"
+    android:padding="16dp"
```
Comment (buracchi) [inline app/src/main/res/layout/menu_exposed_menu_item.xml]: Don't use hardcoded padding

File: app/src/main/java/com/xeroxparc/materialcomponentsdemo/ui/component/menu/MenuActivity.java [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+import android.view.View;
+import android.widget.ArrayAdapter;
+import android.widget.AutoCompleteTextView;
+import android.widget.PopupMenu;
+
+import com.xeroxparc.materialcomponentsdemo.R;
+import com.xeroxparc.materialcomponentsdemo.databinding.ActivityMenuBinding;
+import static com.xeroxparc.materialcomponentsdemo.utils.Utils.inflateSpanTextViewContent;
+
+public class MenuActivity extends AppCompatActivity {
```
Comment (buracchi) [inline app/src/main/java/com/xeroxparc/materialcomponentsdemo/ui/component/menu/MenuActivity.java:17]: Did you mean this?

    public class MenuActivity extends AppCompatActivity {

        @Override
        protected void onCreate(Bundle savedInstanceState) {
            super.onCreate(savedInstanceState);
            Holder holder = new Holder(this);
            setContentView(holder.getRoot());
        }

        @Override
        public boolean onCreateOptionsMenu(Menu menu) {
            MenuInflater inflater = getMenuInflater();
            inflater.inflate(R.menu.menu_overflow_menu, menu);
            return true;
        }

        // popup menu inflater
        public void showPopupMenu(View anchor) {
            PopupMenu popup = new PopupMenu(this, anchor);
            popup.getMenuInflater().inflate(R.menu.menu_dropdown_menu, popup.getMenu());
            popup.show();
        }

        class Holder {

            private ActivityMenuBinding binding;

            Holder(Activity activity) {
                binding = ActivityMenuBinding.inflate(getLayoutInflater());
                binding.appBarContainer.toolbar.setTitle(R.string.menu_title);
                binding.appBarContainer.imageViewBanner.setImageResource(R.drawable.banner_menu);
                inflateSpanTextViewContent(binding, activity);

                binding.appBarContainer.toolbar.inflateMenu(R.menu.menu_overflow_menu);
                new Adapter();
            }
        
            View getRoot() {
                return binding.getRoot();
            }

            // exposed menu adapter
            class Adapter {

                Adapter(){
                    String[] exposedDropdownMenuOptions = new String[] {
                            getString(R.string.menu_headset),
                            getString(R.string.menu_star),
                            getString(R.string.menu_music_note),
                            getString(R.string.menu_info)
                    };
                
                    ArrayAdapter<String> adapter = new ArrayAdapter<>(
                            getApplicationContext(),
                            R.layout.menu_exposed_menu_item,
                            exposedDropdownMenuOptions
                    );
                
                    binding.filledExposedDropdown.setAdapter(adapter);
                }
            }
        }
    }

**Review by buracchi — CHANGES_REQUESTED** [review]

File: app/src/main/java/com/xeroxparc/materialcomponentsdemo/ui/component/menu/MenuActivity.java (hunk 11 above)
Comment (Sph00b) [reply app/src/main/java/com/xeroxparc/materialcomponentsdemo/ui/component/menu/MenuActivity.java:18]: Wait a minute. what the hell "R.layout.menu_exposed_menu_item" should be?

Comment (buracchi) [reply app/src/main/java/com/xeroxparc/materialcomponentsdemo/ui/component/menu/MenuActivity.java:17]: ArrayAdapter<> method is prone to runtime crash

Comment (DaniloDamico) [reply app/src/main/java/com/xeroxparc/materialcomponentsdemo/ui/component/menu/MenuActivity.java:17]: It is however the method suggested by the relative material.io documentation

File: app/src/main/res/values/styles.xml [hunk 12]
```diff
@@ -13,6 +13,9 @@
     <style name="TextAppearance.App.CollapsingToolbar.Expanded" parent="TextAppearance.MaterialComponents.Headline5">
         <item name="android:textColor">@android:color/transparent</item>
         <item name="android:textSize">0sp</item>
+        <item name="android:popupMenuStyle">@null</item>
+        <item name="android:actionOverflowButtonStyle">@android:color/transparent</item>
+
```
Comment (buracchi) [inline app/src/main/res/values/styles.xml:16]: Discard changes, this doesn't seems to affect the issue.

**Review by buracchi — CHANGES_REQUESTED** [review]

**Comment by buracchi** [comment]
The overflow menu should be affected by the Toolbar theme and it should stay hidden while the CollapsingToolbar is expanded.

File: app/src/main/res/values/strings.xml [hunk 13]
```diff
@@ -75,4 +75,69 @@
...
+            <p>
+                Exposed dropdown menus display the currently selected menu item above the menu. They can be used only when a single menu item can be chosen at a time.
+            </p>
+        ]]>
+    </string>
+    <string name="menu_headset">Headset</string>
+    <string name="menu_star">Star</string>
+    <string name="menu_music_note">Music note</string>
+    <string name="menu_info">Info</string>
+    <string name="dropdown_menu">Dropdown Menu</string>
```
Comment (buracchi) [inline app/src/main/res/values/strings.xml:1242]: I'll personally fix it in the release branch

**Review by buracchi — APPROVED** [review]
I'll fix it in the next release

**Pull request merged**